Conditionaly split switch in specialStateTransition method

jsebechlebsky opened the pull request:
Switch in lexer in specialStateTransition can get really big, even after pulling code from case statements into separate methods.

Somewhere slightly over 1500 cases the specialStateTransition method can exceed 64kB limit. 

Implemented transformation will split such a method into several methods with defined maximal number of case statements (1000 by default).

Signed-off-by: Jan Sebechlebsky <jan.sebechlebsky@cz.ibm.com>

szarnekow commented:
Thanks for the PR. In the parser we observe quite often that many many cases have the same body block in those bigger switch statements. Is that also true for the lexer?
Could you please provide a test language grammar that shows the effect of the splitting?

jsebechlebsky commented:
Hi Sebastian. 
I haven't searched for identical bodies in lexer switch statements, I'll try to check it - if that's true, maybe it would a be good place to make some more optimizations. 

I've modified unit test for LexerSpecialStateTransitionSplitter to cover also new functionality. If you take a look at the file there is a result after transformation (Is this sufficient as an example of transformation?). 

While writing the test I realized I probably  misused ignoreCaseCountGuard so I removed check for that. I suppose the purpose of the this variable was just to force LexerSpecialStateTransitionSplitter to pull code from cases into methods regardless of number of cases during the test. 

I've squashed changes on LexerSpecialStateTransitionSplitter into one commit, the test is in the separate commit. If you would like to squash it all into single commit just let me know.

szarnekow commented on tests/org.eclipse.xtext.tests/src/org/eclipse/xtext/generator/parser/splitting/LexerSpecialStateTransitionSplitterTest.java:
```diff
@@ -63,6 +63,43 @@
+			"                        if ( s>=0 ) return s;\n" +
+			"                        break;\n" +
+			"                    case 4 : \n" +
+			"                        int LA12_1 = input.LA(1);\n" +
```
case 4 appears to have the same body as case 2

szarnekow commented:
If your test data was taken from a real lexer, it seems to be the same situtation as for the parser. The case block are identical for different case labels. Do you have the time to work on a mechanism that allows to detect those dups and merge them into a switch with fall-throughs for duplicate case blocks?

jsebechlebsky commented:
You're right, 2 and 4 are semantically identical in this case. I was little worried if the Antlr could generate conditions which would not exclude each other -> in that case the order of the condition testing would matter. But that could be probably easily checked by inspecting Antlr code, or if you are sure about that, I think I can implement the transformation and submit another pull request.

szarnekow commented:
I'm very confident that the order in the if-else checks within the cases doesn't matter. The if-else cascades represent a normalized (sub-)DFA as far as I can tell thus each if-condition describes the conditions for the transition to another state. These transitions cannot overlap each other.
It would be greatly appreciated if you could come up with a logic that determines the identical cases and merges them. Please feel free to add the commits to this PR.

szarnekow commented:
Jan, are you investigating in the discussed optimization?

jsebechlebsky commented:
Hi Sebastian, sorry for a delay, I'm planning to do it on monday.

jsebechlebsky commented:
Hi Sebastian, I haven't finished it yet but so far it seems that on my lexer sample almost 3/4 cases contains identical statements.

szarnekow commented:
That sounds very promising :+1:  Do you plan to support that automated refactoring / code cleanup also for the parser?

jsebechlebsky commented:
Hi Sebastian, It will probably take me some more time to look at it (I probably also need separate permission for this). I would like to ask you - When do you think code from this pull request can be included in the release? We're using automatically generated xtext parsers with custom lexer using semantic predicates which result in non-compilable code in certain cases, I've noticed that quite a lot people using custom lexers are facing similar issue. It would be nice to have this fixed in early Xtext release.

svenefftinge commented:
hi jan, we are going to have a release in the coming days. Tomorrow there will be an RC if there are no stoppers in it we are going to release it next Monday. 

The change looks good, but could potentially introduce problems and I would prefer to have the splitting disabled by default. So if you could provide a PR that completely skips the newly introduced code unless someone sets the new skipping property, I can imagine to have it in 2.9.1, which will be in two or three weeks. Does that sound ok for you?

jsebechlebsky commented:
That sounds great, I'll make the modification and push it to this pull request.

jsebechlebsky commented:
Hi, I've pushed the change - it will turn off special state switch splitting by default (new transformation is completely omitted), it can be explicitly turned on by setting specialStateSwitchSplitting to true.

svenefftinge commented:
I'd prefer a single property that encodes the number and whether it is set or not. 
Because now people could set the number but it wouldn't have any effect if they forgot to enable the splitting, which is unnecessarily complicated. I think either Integer with null or int with a negative value would be a good value for a disabled switch-splitting.

svenefftinge commented:
I merged and did another PR #797

jsebechlebsky commented:
Thanks a lot!
